.github/workflows/kind-e2e.yaml

```diff
@@ -0,0 +1,197 @@
+
+        # Be KinD to these tests.
+        kubectl scale -nknative-serving deployment/chaosduck --replicas=0
+
+    - name: Wait for Webhook to be up
+      working-directory: ./src/knative.dev/serving
+      run: |
+        # We need the webhook to be up
+        kubectl wait pod --for=condition=Ready -n knative-serving  -l app=webhook
+
+    - name: Install kingress provider (Contour)
+      working-directory: ./src/knative.dev/serving
+      if: matrix.kingress == 'contour'
+      run: |
+        set -o pipefail
+
+        # Apply a kingress option.
+        ko resolve -f third_party/contour-latest | \
+          sed 's/LoadBalancer/NodePort/g' | \
+          sed 's/imagePullPolicy:/# DISABLED: imagePullPolicy:/g' | \
+          kubectl apply -f -
+
+        # This tells the tests what namespace to look in for our kingress LB.
+        echo "::set-env name=GATEWAY_OVERRIDE::envoy"
+        echo "::set-env name=GATEWAY_NAMESPACE_OVERRIDE::contour-external"
+
+    - name: Install kingress provider (Kourier)
+      working-directory: ./src/knative.dev/serving
+      if: matrix.kingress == 'kourier'
+      run: |
+        set -o pipefail
+
+        # Apply a kingress option.
+        ko resolve -f third_party/kourier-latest | \
+          sed 's/LoadBalancer/NodePort/g' | \
+          sed 's/imagePullPolicy:/# DISABLED: imagePullPolicy:/g' | \
+          kubectl apply -f -
+
+        # This tells the tests what namespace to look in for our kingress LB.
+        echo "::set-env name=GATEWAY_OVERRIDE::kourier"
+        echo "::set-env name=GATEWAY_NAMESPACE_OVERRIDE::kourier-system"
+
+    - name: Install kingress provider (Istio)
+      working-directory: ./src/knative.dev/serving
+      if: matrix.kingress == 'istio'
+      run: |
+        # TODO(https://github.com/knative-sandbox/net-istio/issues/285): Istio + NodePort
+        exit 1
+
+    - name: Upload Test Images
+      working-directory: ./src/knative.dev/serving
+      run: |
+        # Build and Publish our test images to the docker daemon.
+        ./test/upload-test-images.sh
+
+    - name: Wait for Serving and KIngress to be up
+      working-directory: ./src/knative.dev/serving
+      run: |
+        kubectl wait pod --for=condition=Ready -n knative-serving -l '!job-name'
+        kubectl wait pod --for=condition=Ready -n "${GATEWAY_NAMESPACE_OVERRIDE}" -l '!job-name'
+
+    - name: Run e2e Tests
+      working-directory: ./src/knative.dev/serving
+      run: |
+        set -x
+
+        # Exclude the control-plane node, which doesn't seem to expose the nodeport service.
+        IPS=( $(kubectl get nodes -lkubernetes.io/hostname!=kind-control-plane -ojsonpath='{.items[*].status.addresses[?(@.type=="InternalIP")].address}') )
+
+        # Run the tests tagged as e2e on the KinD cluster.
+        go test -race -count=1 -timeout=20m -tags=e2e ${{ matrix.test-suite }} \
+           --ingressendpoint="${IPS[0]}"
```

Comment:
you may want to lift some of this code to get debug output on failure: https://github.com/knative-sandbox/eventing-rabbitmq/blob/master/.github/workflows/kind-e2e.yaml#L129-L158

Reply:
I think there are enough changes I'd make to that that I'd want to follow up.  Top of mind:
1. Use artifacts instead of catting to stdout (it'd be sweet if we could have separate artifacts for each of your sections)
2. Should we start this (or something like `kail` or `stern`) earlier and upload the artifact on failure to avoid only capturing the tail end of the logs?  Is that even a problem with kind? 🤔

Reply:
It'd actually be sort of sweet if we could encapsulate a KinD action that spun up KinD and if the enclosing test failed uploaded all manner of logs / metrics / ...

Reply:
I did not really get good results with trying to `bg` the `stern` call. Still looking for a valid solution. Best I got was getting logstream to work for real

Reply:
https://github.com/marketplace/actions/kind-kubernetes-in-docker-action
Can this action be used to fulfil the first half of the ask?

Reply:
> It'd actually be sort of sweet if we could encapsulate a KinD action that spun up KinD and if the enclosing test failed uploaded all manner of logs / metrics / ...

Yeah I think to make this real we should make a kind-for-knative action :bikeshed: the name

Reply:
@chaodaiG I've played with some in previous contexts, and this seemed to produce the best result.  IDK that I tried that particular one, and I've also learned a bunch, but I suspect we'll continue refining this until the heat death of the universe :D

Reply:
Another thing we might try is setting up kind in a separate `job` with `kubeconfig` as an output, then we might be able to parallelize things like the image upload and the build (I don't think steps parallelize, only jobs).

But I'd like to checkpoint something that works and than increment towards that glorious future 😅